metrics/otelgrpc/otelgrpc.go

```diff
@@ -0,0 +1,159 @@
+package otelgrpc
+
+import (
+	"context"
+	"strings"
+	"time"
+
+	"github.com/goto/meteor/utils"
+	"go.opentelemetry.io/otel"
+	"go.opentelemetry.io/otel/attribute"
+	"go.opentelemetry.io/otel/metric"
+	semconv "go.opentelemetry.io/otel/semconv/v1.20.0"
+	"google.golang.org/grpc"
+	"google.golang.org/protobuf/proto"
+)
+
+type UnaryParams struct {
+	Start  time.Time
+	Method string
+	Req    any
+	Res    any
+	Err    error
+}
+
+type Monitor struct {
+	duration     metric.Int64Histogram
+	requestSize  metric.Int64Histogram
+	responseSize metric.Int64Histogram
+	attributes   []attribute.KeyValue
+}
+
+func NewOtelGRPCMonitor(hostName string) Monitor {
+	meter := otel.Meter("github.com/goto/meteor/metrics/otelgrpc")
+
+	duration, err := meter.Int64Histogram("rpc.client.duration", metric.WithUnit("ms"))
+	if err != nil {
+		handleOtelErr(err)
+	}
+
+	requestSize, err := meter.Int64Histogram("rpc.client.request.size", metric.WithUnit("By"))
+	if err != nil {
+		handleOtelErr(err)
+	}
+
+	responseSize, err := meter.Int64Histogram("rpc.client.response.size", metric.WithUnit("By"))
+	if err != nil {
+		handleOtelErr(err)
+	}
+
+	return Monitor{
+		duration:     duration,
+		requestSize:  requestSize,
+		responseSize: responseSize,
+		attributes: []attribute.KeyValue{
+			semconv.RPCSystemGRPC,
+			attribute.String("server.address", hostName),
+		},
+	}
+}
+
+func getProtoSize(p any) int {
+	if p == nil {
+		return 0
+	}
+
+	size := proto.Size(p.(proto.Message))
+	return size
+}
+
+func (m *Monitor) RecordUnary(ctx context.Context, p UnaryParams) {
+	reqSize := getProtoSize(p.Req)
+	resSize := getProtoSize(p.Res)
+
+	attrs := m.attributes
+	attrs = append(attrs, attribute.String("rpc.grpc.status_text", utils.StatusText(p.Err)))
+	attrs = append(attrs, parseFullMethod(p.Method)...)
```

Comment:
Would this lead to concurrent modification of the global slice?

Reply:
should not, because we're creating new var in line 74 and this is not a map data type. Let me add test cases to make sure

Reply:
A slice comprises of an underlying array that gets modified or replaced when we call append. Concerned about concurrent modification of the array

Reply:
updated